### PR TITLE
Filter Kubernetes Nodes on multiple fields

### PR DESCRIPTION
**Pull request by dotNomad**

### CODE REVIEW

**Comment by dotNomad** [comment]
Added a list of filters to the `IListConfig` interface that contains all of the fields we want to filter on. This creates a selection field that changes what the filter string filters on.
The transformEntities of the `kubernetes-nodes-list-config.service.ts` consumes the `filterKey` (the id of the filter we are using) to decide how to filter the Data Source.

**Comment by jeremy-moffitt** [comment]
looks like either something is missing or the unit tests need updated:

> Property 'getFilters' is missing in type '{ allowSelection: false; cardComponent: null; defaultView: ListView; enableTextFilter: false; getColumns: () => any; getDataSource: () => any; getGlobalActions: () => any; getInitialised: () => any; ... 6 more ...; viewType: ListViewTypes.BOTH; }'

**Review by GarySmith — COMMENTED** [review]
We talked about making `getFilters` optional, which would avoid requiring adding the empty method to all of these places. Is that still a possibility?

**Comment by dotNomad** [comment]
I did attempt to make the `getFilters` optional, but ran into some other issues that would have forced type definition changes or change implements / extends on classes. Keeping `getFilters` felt like the less impactful change.

File: custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts [hunk 1]
```diff
@@ -68,28 +76,79 @@ export class KubernetesNodesListConfigService implements IListConfig<KubernetesN
...
   constructor(
     store: Store<AppState>,
     kubeId: BaseKubeGuid,
   ) {
-    this.nodesDataSource = new KubernetesNodesDataSource(store, kubeId, this);
-  }
+    const transformEntities: DataFunction<KubernetesNode>[] = [
+      (entities: KubernetesNode[], paginationState: PaginationEntityState) => {
+        const filterKey = paginationState.clientPagination.filter.filterKey;
+        const filterString = paginationState.clientPagination.filter.string.toUpperCase();
```
Comment (richard-cox) [inline custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts:131]: If `filterString` is a falsy can return `entities` without continuing

File: custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts [hunk 2]
```diff
@@ -68,28 +76,79 @@ export class KubernetesNodesListConfigService implements IListConfig<KubernetesN
...
+
+        switch (filterKey) {
+          case KubernetesNodesListFilterKeys.IP_ADDRESS:
+            return entities.filter(node => {
+              const internalIP: string = node.status.addresses.find(address => {
```
Comment (richard-cox) [inline custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts]: I'm not up to speed with kube spec, but would it be possible to have multiple or no addresses marked at `InternalIP`. If so would be worth covering both cases just in case

File: custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts (hunk 1 above)
Comment (richard-cox) [inline custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts:131]: To avoid converting each property/properties in entities to uppercase each time we pass through transform the filterString could be a regex that ignores case. So something like...
`const filterRegEx = new RegExp(filterString, "i")`
`<x>.search(filterRegEx)`
We probably haven't done this in other places when we should have.

File: src/frontend/packages/store/src/reducers/pagination-reducer/pagination-reducer-set-client-filter-key.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    const clientPagination = spreadClientPagination(state.clientPagination);
+
+    return {
+        ...state,
+        error: false,
+        clientPagination: {
+            ...clientPagination,
+            filter: {
+                ...clientPagination.filter,
+                items: {
```
Comment (richard-cox) [inline src/frontend/packages/store/src/reducers/pagination-reducer/pagination-reducer-set-client-filter-key.ts]: Think the `items` spread can go

File: src/frontend/packages/core/src/shared/components/list/list.component.html [hunk 4]
```diff
@@ -201,4 +215,11 @@
       </ng-container>
     </div>
   </ng-container>
-</div>
\ No newline at end of file
+</div>
+
+<ng-template #refreshButton>
```
Comment (richard-cox) [inline src/frontend/packages/core/src/shared/components/list/list.component.html]: Probably a merge artefact, can be deleted

**Review by richard-cox — CHANGES_REQUESTED** [review]
I've done a review of the PR, overall looking good! There's some niggly things inline. Additionally....
- We've spoken to Sorin and raised a couple of things to look at
  - The position of the `Filter Selection` text
  - General pattern of filtering by data that's not shown in the table (ip address & labels)
- ~At some point it would be nice to port this upstream, so limiting the changes such as getFilters would be good. I'll look into this and see if there's a clean way of making it optional~
- ~I need to come back and test some more edge case like scenarios~
- ~It looks like the pagination e2e tests are failing due to the additional mat select in the dom~
  - see `application-wall-e2e.spec.ts` createPageSizeSelectId
  - The `mat-select-<x>` will probably need to be tweaked, a better solution would be better selectors but for now lets low bar
  - I'll send Instructions on how to run e2e tests in RC

File: custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts (hunk 2 above)
Comment (dotNomad) [reply custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts]: It looks like its expected to only be a string, not a list of IPs, and documentation only references a single `InternalIP`
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#nodeaddress-v1-core

It does look like it could be `ExternalIP` or `InternalIP` though so we could have a separate filter for either and handle it better if one isn't present.

File: src/frontend/packages/core/src/shared/components/list/list.component.html (hunk 4 above)
Comment (dotNomad) [reply src/frontend/packages/core/src/shared/components/list/list.component.html]: Good catch.

File: custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts (hunk 1 above)
Comment (dotNomad) [reply custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts:131]: Good change to clean up the code, but does add regex support to the filtering which could be confusing if not apparent to the user. I'm initially thinking about IP Addresses where the `.` would the regex dot.
We could escape the filterString so the `.` would actually filter for `.` (and any other special regex characters), but that just adds back in code complexity.

File: custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts (hunk 2 above)
Comment (richard-cox) [reply custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts]: I've pushed a change that will fall back on the `ExternalIP` and also handle the case where neither are found. If the `ExternalIP` check goes against the spec of this feature or you'd rather it not there just let me know and will pull it out.

File: custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts (hunk 1 above)
Comment (richard-cox) [reply custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts:131]: Good point! Lets stick with what's there.

**Comment by richard-cox** [comment]
We've had a round of review with Sorin and would like to address the following points
- Present information in the table that is being searched on
  - Add `IP` column with globe icon and ip address in pop over
  - Add `Labels` column containing an `app-chips` component. This will show chips that fit into the space with an option to show all
  - Combine `Disk Pressure` and `Memory Pressure` columns into one
  - See examples in https://xd.adobe.com/spec/dcd4203c-d334-462e-7921-ea6c3ef41151-aae5/screen/a2ea0108-1ae5-418f-8153-bc2a1fe985d5/Filter/ for details
- Retain the `Filter selection` for now and look for user input
